### ISSUE TITLE
Bug: MREVERSE flag not set

### ISSUE THREAD

**Issue by nathanhaigh** [csiro-crop-informatics/biokanga#12]
I don't think BioKanga always correctly sets the MREVERSE SAM flag (`0x20`). It may only affect read pairs from small inserts such that the start/end coordinates are the same for both reads in a pair.

I think there is an issue when READ1 has REVERSE (`0x10`) set but then READ2 does not get MREVERSE (`0x20`) set. An example is this:

```
D00615:71:CC1WCANXX:2:2301:2669:90742   131     chr6D_part1     100001827       255     80M     =       100001827       80
D00615:71:CC1WCANXX:2:2301:2669:90742   83      chr6D_part1     100001827       255     80M     =       100001827       80
```

I think the relevant code resides here: https://github.com/csiro-crop-informatics/biokanga/blob/ff73bf65fd17f860b1ee568685bd0ef73b196950/biokanga/Aligner.cpp#L5656 and involves `cSAMFlgAS` and `cSAMFlgMateAS`

**Comment by nathanhaigh** [comment]
I should also add, that if READ2 has REVERSE (`0x10`) set, it appears that READ1 does get MREVERSE (`0x20`) set correctly.

**Comment by nathanhaigh** [comment]
ping @rsuchecki

**Comment by kit4b** [comment]
Thanks for the notification , I’ll look into it.

**Comment by kit4b** [comment]
"I don't think BioKanga always correctly sets the MREVERSE SAM flag (0x20). It may only affect read pairs from small inserts such that the start/end coordinates are the same for both reads in a pair.
I think there is an issue when READ1 has REVERSE (0x10) set but then READ2 does not get MREVERSE (0x20) set. An example is this:
D00615:71:CC1WCANXX:2:2301:2669:90742   131     chr6D_part1     100001827       255     80M     =       100001827       80
D00615:71:CC1WCANXX:2:2301:2669:90742   83      chr6D_part1     100001827       255     80M     =       100001827       80"

The flags in field 2 of the SAM alignment extraction map to hexadecimal as being -
131 => 0x83 or in binary 10000011
83 =>  0x53 or in binary  01010011

From the SAM V1.5 format specification -
"2. FLAG: Combination of bitwise FLAGs. Each bit is explained in the following table:
Bit Description 
1 0x1 template having multiple segments in sequencing 
2 0x2 each segment properly aligned according to the aligner 
4 0x4 segment unmapped 
8 0x8 next segment in the template unmapped 
16 0x10 SEQ being reverse complemented 
32 0x20 SEQ of the next segment in the template being reverse complemented 
64 0x40 the first segment in the template 
128 0x80 the last segment in the template 
256 0x100 secondary alignment 
512 0x200 not passing filters, such as platform/vendor quality controls 
1024 0x400 PCR or optical duplicate 
2048 0x800 supplementary alignment"

The read 'D00615:71:CC1WCANXX:2:2301:2669:90742' has flags showing this read as:
a) Part of a template (PE in this instance) with multiple segments (2 reads) with each read of the PE aligned
b) This segment (read) was not reverse complemented
c) The next segment (read) is flagged as not reverse complemented but this flag is not relevant as there  is no next segment in the current template - this is the last segment (read)! 
d) This segment (read) is the last segment in the template

And the read 'D00615:71:CC1WCANXX:2:2301:2669:90742' has flags showing this read as:
a) Part of a template with multiple segments with each segment (2 reads) of the PE aligned
b) This segment (read) was reverse complemented
c) The next segment (read) was not reverse complemented 
d) This segment (read) is the first segment in the template

The flags are consistent, there is a PE read marked as being the first which was reverse complemented
in order to be aligned and flags for this read show that there is another subsequent read in the template
which was not reverse complemented. 
The subsequent read in the template is marked as being the last and was not reverse complemented. 
There is no following read in the template and hence bit 5 (0x20) is not set. 
WRT "I think there is an issue when READ1 has REVERSE (0x10) set but then READ2 does not get MREVERSE (0x20) set", the only 'wraparound' requirement is for the last read in a template to reference the first read alignment through the read alignment 'RNEXT' field. Flag bit 0x20 in the read marked as last has no SAM v1.5 mandated requirement to reflect the alignment orientation of the first read in that template through the flags.

SAM v1.5 imposes no ordering requirements on alignments within a SAM format conformant file. BioKanga does sort by ascending targetseq.loci before reporting it's alignments in SAM format to aid post-processing of alignments by other applications.

Any application using SAM as input should be prepared to accept unordered alignments, unordered template segments, flag bit 0x20 having no significance if flag bit 0x80 is set, and segments within a template having unexpected alignment orientations.

**Comment by rsuchecki** [comment]
### As per SAM format v1.5 

7.RNEXT: Reference sequence name of the primary alignment of the NEXT read in the template.  **For the last read, the next read is the first read in the template**  

### Also note: 

If RNEXT is ‘*’, no assumptions can be made on PNEXT and bit 0x20

### My understanding/expectation

Although not explicitly stated, for consistency I would expect the notion of *first read in the template being the next read for the last read in the template* to apply to flags as well

**Comment by kit4b** [comment]
Subsequent to internal discussions I decided to set flags bit 0x20 for the last segment in the template according to the orientation of the first segment in the same template. I considered the side effects this may have on existing processing pipelines and concluded that robust downstream application processing of the SAM/BAM alignments would be ignoring this flag on last segments e.g. negligible risks of side effects, whilst less robust processing implementations with dependencies on the flag may be able to utilise this additional relative orientation information.

But when checking the existing code, I determined that functionality already exists for flag bit 0x20 being set (I must have considered this issue some years ago) on the last segment to the orientation of the first , and this is evidenced in the following two PE alignment examples.  

SRR3508141.23016401.2	163	Chr1	4064	255	100M	=	4159	195	AATCGACAGAGCAGGACAACGGTTTCTGGTAAATGGAAGCTTACCGGAGAATCTGTTGAGGTCAAGGACCAGAGGGGATTTTGTAGTGAGGGCTTTCGTG	*
SRR3508141.23016401.1	83	Chr1	4159	255	100M	=	4064	195	TCGTGGTAAGATTGGTCATAAAAGGGTTTTGGTGTTCCTCGATGGAAGATACCCTGACAAAACCAAATCTGATTGGGTTATCCACGAGTTCCACTACGAC	*

SRR3508141.33047851.1	99	Chr1	6899	255	100M	=	6931	132	CTGGAAGCTGATGGGATCATACGATTCTGAAAAAATAAGACATATATTGCAACAGAGATCCAATTTGTATCAAAAATATTGTCGGCTCAAAAATCTGACC	*
SRR3508141.33047851.2	147	Chr1	6931	255	100M	=	6899	132	AAATAAGACATATATTGCAACAGAGATCCAATTTGTATCAAAAATATTGTCGGCTCAAAAATCTGACCCACCAAGAATCTAATCAAGTGCGCGATTAAGC	*

83  =>  0x53      0101 0011  first segment, segment was reverse complemented, next segment was not reverse complemented 
163 => 0xA3     1010 0011  last segment, segment was not reverse complemented, next segment was reverse complemented
99  =>  0x63     0110 0011 first segment, segment was not reverse complemented, next segment was reverse complemented
147=> 0x93     1001 0011  last segment, segment was reverse complemented, next segment was not reverse complemented

There may be some edge case whereby bit 0x20 is not being set, but without knowledge of the biokanga parameterisation used, targets, logs, and sample readsets, then these potential edge cases can’t be exercised and code paths traced.
